File: backend/src/main/java/com/ufc/backend/backend/services/CourseService.java
package com.ufc.backend.backend.services;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.ufc.backend.backend.exceptions.CouldNotLoadTheClassesException;
import com.ufc.backend.backend.exceptions.ObjectNotFoundException;
import com.ufc.backend.backend.model.Classes;
import com.ufc.backend.backend.model.Course;
import com.ufc.backend.backend.repositories.ClassRepository;
import com.ufc.backend.backend.repositories.CourseRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.io.File;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class CourseService {

    @Autowired
    private CourseRepository courseRepository;
    @Autowired
    private ClassRepository classRepository;


    public void populateCourses() {
        try {
            File in = new File(Objects.requireNonNull(this.getClass().getClassLoader().getResource("courses/cc.json")).getFile());
            ObjectMapper mapper = new ObjectMapper();
            // Necessary to not fail when accessing the properties : ref, id. for MongoDB relation
            mapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
            Course course = mapper.readValue(in, Course.class);
            classRepository.saveAll(course.getMandatoryClasses());
            courseRepository.save(mapper.readValue(in, Course.class));
        } catch (Exception e) {
            throw new CouldNotLoadTheClassesException();
        }
    }

    public Course findById(String id) {
        return courseRepository.findById(id).orElseThrow(() -> new ObjectNotFoundException(id));
    }

    public List<Course> findAll() {
        return courseRepository.findAll();
    }

    public List<Classes> findMandatory(String id) {
        return findById(id).getMandatoryClasses();
    }

    public List<Classes> findOptional(String id) {
        return findById(id).getOptionalClasses();
    }

    public List<Classes> findBySemester(String id, Integer semester) {
        return findById(id).getMandatoryClasses().stream().filter(obj -> Objects.equals(obj.getSemester(), semester)).collect(Collectors.toList());
    }

}

File: backend/src/main/java/com/ufc/backend/backend/services/CourseService.kt
package com.ufc.backend.backend.services

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.ufc.backend.backend.exceptions.ObjectNotFoundException
import com.ufc.backend.backend.model.subject.Subject
import com.ufc.backend.backend.model.course.Course
import com.ufc.backend.backend.repositories.SubjectRepository
import com.ufc.backend.backend.repositories.CourseRepository
import org.springframework.stereotype.Service
import java.io.File
import java.util.*

@Service
class CourseService(
    private val courseRepository: CourseRepository,
    private val subjectRepository: SubjectRepository
) {

    fun populateCourses() {
        try {
            val file = File(Objects.requireNonNull(this.javaClass.classLoader.getResource("courses/cc.json")).file)
            val mapper = ObjectMapper()
            mapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
            val course = mapper.readValue(file, Course::class.java)
            subjectRepository.saveAll(course.mandatorySubjects)
            courseRepository.save(mapper.readValue(file, Course::class.java))
        } catch (e: Exception) {
            throw e
        }
    }

    fun findById(id: String): Course {
        return courseRepository.findById(id).orElseThrow { ObjectNotFoundException(id) }!!
    }

    fun findAll(): Collection<Course> {
        return courseRepository.findAll()
    }

    fun findMandatory(id: String): Collection<Subject> {
        return findById(id).mandatorySubjects
    }

    fun findBySemester(id: String, semester: Int): Collection<Subject> {
        return findById(id).mandatorySubjects.filter { it.semesterIndex == semester }
    }
}